src/ripple/app/misc/impl/AmendmentTable.cpp

```diff
@@ -489,8 +492,10 @@ AmendmentTableImpl::doVoting(
         }
     }
 
-    vote->mThreshold =
-        std::max(1, (vote->mTrustedValidations * majorityFraction_) / 256);
+    vote->mThreshold = rules.enabled(ripple::fix3396)
+        ? std::max(1, (vote->mTrustedValidations * majorityFraction_.new_) / 10)
```

Comment:
I appreciate the well-thought out comment (and agree with the 8 -> 80 change), but I find the division code confusing to read. Once I read it carefully, I understand what's happening, but I'm not confident it behaves _identically_ pre-amendment. I am hoping we can do something even better:

We should push all this logic into `AmendmentSet` and reduce it's public interface to a _single_ function: `bool passes(uint256) const`:

```
struct AmendmentSet
{
private:
    // How many yes votes each amendment received
    hash_map<uint256, int> votes_;

    // number of trusted validations
    int mTrustedValidations = 0;

    // number of votes needed
    int mThreshold;

    Rules const& rules;

public:
    AmendmentSet(
        Rules const& rules,
        MajorityFraction const& majorityFraction,
        std::vector<std::shared_ptr<STValidation>> const& valSet)
    {
        for (auto const& val : valSet)
        {
            if (val->isTrusted())
            {
                std::set<uint256> ballot;

                if (val->isFieldPresent(sfAmendments))
                {
                    auto const choices = val->getFieldV256(sfAmendments);
                    ballot.insert(choices.begin(), choices.end());
                }

                ++mTrustedValidations;

                for (auto const& amendment : ballot)
                    ++votes_[amendment];
            }
        }

        mThreshold = rules.enabled(fix3396)
            ? std::max(1, (vote->mTrustedValidations * majorityFraction.pre3396.num) / majorityFraction.pre3396.den)
             : std::max(1, (vote->mTrustedValidations * majorityFraction.post3396.num) / majorityFraction.post3396.den);
    }

    bool
    passes(uint256 const& amendment) const
    {
        auto const& it = votes_.find(amendment);

        if (it == votes_.end())
            return false;

        // Before this fix, it was possible for an amendment to activate with a percentage
        // slightly less than 80% because we compared for "greater than or equal to"
        // instead of strictly "greater than".
        if (!rules.enabled(fix3396))
            return it->second >= fix3396;

        return it->second > fix3396;
    }
};
```

In reviewing this code, I spotted something else that we may want to include: we "import" the `std::vector` of validations into a `std::set` to "de-duplicate" and make it easier to search.

I propose, instead, that, if this amendment is enabled, we skip that and instead call `std::sort` and `std::unique` and if a duplicate vote is present, that validator's vote for this round should be discarded in toto, and a warning should be produced.

I'll code it up.